File: insertionSort/insertionSort.cpp
#include <bits/stdc++.h>
using namespace std;

//insertion sort in C++
template <typename T,
            typename Comp = std::less<T> >
void insertionSort(vector<T> &arr, Comp mLess = Comp()){
    for(int step = 1; step < (int)arr.size(); ++step){
        T value = arr[step];
        int j = step -1;

        //Compare each value with comparator
        //Compare to each element on the left until an element is less than the value
        while(j >= 0 && mLess(value, arr[j])){
            arr[j +1] = arr[j];
            --j;
        }
        arr[j +1] = value;
    }
    return;
}

File: insertionSort/insertionSort.py
#insertion sort in python
def insertionSort(array, mLess = lambda x, y : x < y):

	for step in range(1, len(array)):
		value = array[step]
		j = step -1

		#compare value with each element on the left of it until it is less than an element
		while j >= 0 and mLess(value, array[j]):
			array[j +1] = array[j]
			j -= 1
		
		#place the value after the element just less than it
		array[j +1] = value
	return